WordPress/wordpress-develop
# 55677 Avoid duplicated query when retrieving empty collection from WP REST

## The issue

**WP REST API** has this logic to re-run the query without the page parameter if we didn't get any posts back from the main one. This is to handle the case where we requested an extra page of a collection that is empty. In that case, we want the API to return an empty collection, but still return the total number of items in that collection. Therefore that extra query gives that total back.

However, it currently doesn't check if we were on the first page already. 
In that specific case, the second query would be the exact same as the first one and therefore wouldn't provide any additional value (the total of items would still be zero). This is why we sometimes ends up with a duplicated query.
By checking if we're already on the first page, we can avoid a duplicated query for empty collections.

Note that the trac ticket refers to the error happening on attachments endpoint, however any empty collection of posts will be affected (This is why some tests that were previously checking for two queries and had a "FIXME" comment, have been updated to one).

## How to test

- **Query Monitor** plugin should be enabled to detect any duplicate queries. 
- Open the gutenberg editor
- Open the JS console
- Type `await wp.apiFetch({ path: 'wp/v2/media?_envelope&media_type=text'}`
- The object returned should contain the number of duplicated queries. On trunc, there will be a duplicate which this patch will fix.


Trac ticket: [https://core.trac.wordpress.org/ticket/55677](https://core.trac.wordpress.org/ticket/55677)


---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**
